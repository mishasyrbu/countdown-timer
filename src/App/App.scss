.app-container {
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  > div:not(:last-child) {
    margin-bottom: 26px;
  }
}
