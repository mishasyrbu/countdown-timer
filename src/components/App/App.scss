.app-container {
  height: 100vh;
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  > div:not(:last-child) {
    margin-bottom: 26px;
  }
}
