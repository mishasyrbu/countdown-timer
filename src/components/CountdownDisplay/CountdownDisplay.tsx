import React, {useEffect, useState} from 'react';
import classNames from 'classnames';

import { CountdownDisplayProps } from '.';
import styles from './CountdownDisplay.module.scss';

const CountdownDisplay = (props: CountdownDisplayProps) => {
    const { className, isOn = false, onPlayStop, initialTime, time } = props;
    const [isFirstRun, setIsFirstRun] = useState<boolean>(true);
    const classes = classNames(styles.countdownDisplay, className);
    const minutes = Math.floor(time / 60);
    const seconds = time - minutes * 60;

    const onPlayPauseClick = (event: React.MouseEvent<HTMLButtonElement>) => {
        event.preventDefault();

        time > 0 && onPlayStop();
    };

    const renderInfoBox = () => {
        const timesUp = time === 0;
        const lessThanHalfLeft = !timesUp && initialTime && initialTime / 2 > time;
        const classes = classNames(styles.messageInfo, {
            [styles.redText]: !timesUp && time <= 20,
            [styles.blink]: !timesUp && time <= 10,
        });

        return (
            <div className={classes}>
                {!isFirstRun && lessThanHalfLeft && <i>More than halfway there!</i>}
                {!isFirstRun && timesUp && <i>Timeâ€™s up!</i>}
            </div>
        );
    };

    useEffect(() => {
        if (time && isFirstRun) {
            setIsFirstRun(false);
        }
    }, [time, isFirstRun]);

    return (
        <div className={classes}>
            {renderInfoBox()}
            <div className={styles.countdown}>
                {minutes < 10 ? `0${minutes}` : minutes}:{seconds < 10 ? `0${seconds}` : seconds}
            </div>
            {!!time && (
                <button className={styles.playPause} onClick={onPlayPauseClick}>
                    <div className={isOn ? styles.pause : styles.play} />
                </button>
            )}
        </div>
    );
};

export default CountdownDisplay;
