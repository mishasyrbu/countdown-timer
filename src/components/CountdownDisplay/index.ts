export { default } from './CountdownDisplay';
export * from './CountdownDisplay.model';
