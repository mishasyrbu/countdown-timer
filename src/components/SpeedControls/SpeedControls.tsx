import React, { useState } from 'react';
import classNames from 'classnames';

import Button from '../Button';
import { TIMER_SPEED_MODE, SpeedControlsProps, SpeedModeType } from '.';
import styles from './SpeedControls.module.scss';

const SpeedControls = (props: SpeedControlsProps) => {
    const { className, onChange } = props;
    const classes = classNames(styles.speedControls, className);
    const [selectedMode, setSelectedMode] = useState<SpeedModeType>(TIMER_SPEED_MODE.X1);

    const onButtonClick = (mode: SpeedModeType) => () => {
        setSelectedMode(mode);
        onChange(mode.value);
    };

    return (
        <div className={classes}>
            {Object.values(TIMER_SPEED_MODE).map((mode) => (
                <Button
                    key={mode.name}
                    className={styles.button}
                    type="action"
                    onClick={onButtonClick(mode)}
                    isActive={selectedMode.name === mode.name}
                >
                    {mode.name}
                </Button>
            ))}
        </div>
    );
};

export default React.memo(SpeedControls);
