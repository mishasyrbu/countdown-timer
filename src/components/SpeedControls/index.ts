export { default } from './SpeedControls';
export * from './SpeedControls.model';
export * from './SpeedControls.const';
