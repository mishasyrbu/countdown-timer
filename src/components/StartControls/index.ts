export { default } from './StartControls';
export * from './StartControls.model';
