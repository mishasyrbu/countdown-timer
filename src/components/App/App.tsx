import React from 'react';

import { AppProps, AppState } from '.';
import SpeedControls from '../SpeedControls';
import CountdownDisplay from '../CountdownDisplay';
import StartControls from '../StartControls';
import './App.scss';

class App extends React.Component<AppProps, AppState> {
    private timer: NodeJS.Timeout | null = null;
    private audioNotificationRef = React.createRef<HTMLMediaElement>();

    state = {
        speed: 1000,
        isOn: false,
        initialTime: 0,
        time: 0,
    };

    startTimer = async () => {
        const { speed } = this.state;

        await this.setState({ isOn: true });
        this.timer = setInterval(() => {
            this.setState(({ time }): AppState => {
                if (time === 0) {
                    this.timer && clearInterval(this.timer);
                    this.playNotification();

                    return { time, initialTime: null, isOn: false };
                }

                return { time: time - 1 };
            });
        }, speed);
    };

    stopTimer = async () => {
        await this.setState({ isOn: false });
        this.timer && clearInterval(this.timer)
    };

    playNotification = () => {
        this.audioNotificationRef
        && this.audioNotificationRef.current
        && this.audioNotificationRef.current.play();
    };

    onStartClick = async (minutes: number) => {
        const seconds = minutes * 60;

        await this.stopTimer();
        await this.setState({ initialTime: seconds, time: seconds });
        this.startTimer();
    };

    onPlayStopClick = () => {
        const { isOn } = this.state;

        if (isOn) {
            return this.stopTimer();
        }

        return this.startTimer();
    };

    onSpeedChange = async (value: number) => {
        const { isOn } = this.state;

        await this.setState({ speed: 1000 / value });

        if (isOn) {
            await this.stopTimer();
            await this.startTimer();
        }
    };

    render() {
        const { isOn, initialTime, time } = this.state;

        return (
            <div className="app-container">
                <StartControls onStart={this.onStartClick} />
                    <CountdownDisplay
                        isOn={isOn}
                        time={time}
                        initialTime={initialTime}
                        onPlayStop={this.onPlayStopClick}
                    />
                <SpeedControls onChange={this.onSpeedChange} />
                <audio ref={this.audioNotificationRef}>
                    <source src={require('./assets/audio/notification.ogg')} type="audio/ogg" />
                    <source src={require('./assets/audio/notification.mp3')} type="audio/mpeg" />
                </audio>
            </div>
        );
    }
}

export default App;
