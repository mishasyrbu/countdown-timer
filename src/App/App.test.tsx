import React from 'react';
import { shallow, mount } from 'enzyme';

import App from './App';
import StartControls from '../components/StartControls';
import SpeedControls from '../components/SpeedControls';
import CountdownDisplay from '../components/CountdownDisplay';

describe('<App />', () => {
    it('should match the snapshot', () => {
        const app = shallow(<App />);

        expect(app.html()).toMatchSnapshot();
    });

    it('should contain all main components', () => {
        const app = shallow(<App />);

        expect(app.find(StartControls).length).toEqual(1);
        expect(app.find(CountdownDisplay).length).toEqual(1);
        expect(app.find(SpeedControls).length).toEqual(1);
    });

    it('should set all values in state on timer start', async () => {
        const app = mount(<App />);
        const appInstance = app.instance() as App;

        await appInstance.onStartClick(100);

        expect(appInstance.timer).toBeTruthy();
        expect(appInstance.state.isOn).toBeTruthy();
        expect(appInstance.state.time).toEqual(6000);
        expect(appInstance.state.initialTime).toEqual(6000);
    });

    it('should properly play/stop', async () => {
        const app = mount(<App />);
        const appInstance = app.instance() as App;

        await appInstance.onStartClick(100);

        expect(appInstance.state.isOn).toBeTruthy();

        await appInstance.onPlayStopClick();

        expect(appInstance.state.isOn).toBeFalsy();

        await appInstance.onPlayStopClick();

        expect(appInstance.state.isOn).toBeTruthy();
    });

    it('should properly speed change', async () => {
        const app = mount(<App />);
        const appInstance = app.instance() as App;

        // check initial value
        expect(appInstance.state.speed).toEqual(1000);

        await appInstance.onSpeedChange(2);

        expect(appInstance.state.speed).toEqual(500);
    });

    it('should properly speed change when timer is on', async () => {
        const app = mount(<App />);
        const appInstance = app.instance() as App;

        await appInstance.onStartClick(100);
        await appInstance.onSpeedChange(2);

        expect(appInstance.state.speed).toEqual(500);
    });

    it('should play notification sound', async () => {
        const playStub = jest
            .spyOn(window.HTMLMediaElement.prototype, 'play')
            .mockImplementation(() => Promise.resolve());
        const app = mount(<App />);
        const appInstance = app.instance() as App;

        await appInstance.playNotification();

        expect(playStub).toHaveBeenCalled();
        playStub.mockRestore();
    });

    it('should start timer with correct arguments', async () => {
        const playStub = jest
            .spyOn(window.HTMLMediaElement.prototype, 'play')
            .mockImplementation(() => Promise.resolve());
        const app = mount(<App />);
        const appInstance = app.instance() as App;
        appInstance.setState({ time: 1, initialTime: 1, isOn: false });

        await appInstance.startTimer();

        expect(appInstance.state.isOn).toBeTruthy();

        await new Promise((r) => setTimeout(r, 3000));

        expect(appInstance.state.isOn).toBeFalsy();
        expect(appInstance.state.time).toBeFalsy();
        expect(appInstance.state.initialTime).toBeFalsy();
        expect(playStub).toHaveBeenCalled();
        playStub.mockRestore();
    });
});
