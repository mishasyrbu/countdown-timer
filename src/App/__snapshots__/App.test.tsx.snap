// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match the snapshot 1`] = `"<div class=\\"app-container\\"><div class=\\"startControls\\"><span class=\\"title\\">Countdown:</span><input type=\\"text\\" class=\\"input\\" placeholder=\\"(Min)\\" pattern=\\"[0-9]*\\" value=\\"\\"/><button class=\\"button disabled primary\\">START</button></div><div class=\\"countdownDisplay\\"><div class=\\"messageInfo\\"></div><div class=\\"countdown\\">00:00</div></div><div class=\\"speedControls\\"><button class=\\"button active button action\\">1X</button><button class=\\"button button action\\">1.5X</button><button class=\\"button button action\\">2X</button></div></div>"`;
